### PR TITLE
[MIPS] LLVM data layout give i128 an alignment of 16 for mips64

### PR DESCRIPTION
**Pull request by yingopq**
Fix parts of #102783.

### CODE REVIEW

File: llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp [hunk 1]
```diff
@@ -75,6 +75,14 @@ TEST(DataLayoutUpgradeTest, ValidDataLayoutUpgrade) {
   EXPECT_EQ(UpgradeDataLayoutString("E-m:e-i64:64-n32:64-S128", "sparcv9"),
             "E-m:e-i64:64-i128:128-n32:64-S128");
 
+  // Check that SPARC targets add -i128:128.
```
Comment (brad0) [inline llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp:78]: Should say MIPS.

Comment (yingopq) [reply llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp]: Sorry for my mistake, updated.

**Review by nikic — COMMENTED** [review]
This looks technically fine, but I'm having trouble finding a normative reference for this. Is there an ABI specification for N32/N64 somewhere?

Also, what alignment does clang use for o32 with ForceEnableInt128? https://github.com/llvm/llvm-project/blob/cb2f1619575400a402ffe11040bb927b0fa35910/clang/lib/Basic/Targets/Mips.h#L441 Possibly we need to specify the alignment in that case as well to cover that option.

**Comment by yingopq** [comment]
> This looks technically fine, but I'm having trouble finding a normative reference for this. Is there an ABI specification for N32/N64 somewhere?
> 
> Also, what alignment does clang use for o32 with ForceEnableInt128?
> 
> https://github.com/llvm/llvm-project/blob/cb2f1619575400a402ffe11040bb927b0fa35910/clang/lib/Basic/Targets/Mips.h#L441
> 
> Possibly we need to specify the alignment in that case as well to cover that option.

I did not find any ABI specification for N32/N64 about i128 alignment, and I am doing research about the ForceEnableInt128.

**Comment by yingopq** [comment]
> Also, what alignment does clang use for o32 with ForceEnableInt128

I did test about o32 with ForceEnableInt128:
```
$ cat 1.c
#include <stdio.h>

int main()
{
    printf("alignment of int128: %zd\n", _Alignof(__int128));
    return 0;
}

$ sudo ./build/bin/clang -target mipsel-unknown-linux-gnu -fforce-enable-int128 -emit-llvm -S 1.c
$ sudo ./build/bin/llc -march=mipsel -target-abi o32 -relocation-model=pic 1.ll
$ sudo ./build/bin/clang -target mipsel-unknown-linux-gnu -fPIC -pthread 1.s -o a1main

root@debian-sid-mipsel:~# ./a1main 
alignment of int128: 16

root@debian-sid-mipsel:~# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/mipsel-linux-gnu/12/lto-wrapper
Target: mipsel-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 12.2.0-14' --with-bugurl=file:///usr/share/doc/gcc-12/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-12 --program-prefix=mipsel-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libitm --disable-libsanitizer --disable-libquadmath --disable-libquadmath-support --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-multilib --with-arch-32=mips32r2 --with-fp-32=xx --with-madd4=no --with-lxc1-sxc1=no --enable-targets=all --with-arch-64=mips64r2 --enable-checking=release --build=mipsel-linux-gnu --host=mipsel-linux-gnu --target=mipsel-linux-gnu
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.2.0 (Debian 12.2.0-14) 

```
> Possibly we need to specify the alignment in that case as well to cover that option.

Thanks for your reminder, I would add this.

**Comment by yingopq** [comment]
> This looks technically fine, but I'm having trouble finding a normative reference for this. Is there an ABI specification for N32/N64 somewhere?
> 
> Also, what alignment does clang use for o32 with ForceEnableInt128?
> 
> https://github.com/llvm/llvm-project/blob/cb2f1619575400a402ffe11040bb927b0fa35910/clang/lib/Basic/Targets/Mips.h#L441
> 
> Possibly we need to specify the alignment in that case as well to cover that option.

@nikic I checked several other arch-32 situations with ForceEnableInt128, and they all use `alloca i128, align 16`.
But they did not modify the target datalayout content with `i128:128`, should mips be consistent?

**Comment by nikic** [comment]
> @nikic I checked several other arch-32 situations with ForceEnableInt128, and they all use `alloca i128, align 16`. But they did not modify the target datalayout content with `i128:128`, should mips be consistent?

I think either way is fine here. Something to consider though, is that the AutoUpgrade for the DataLayout has to match. So if you don't add `i128:128` to the o32 DL, then it shouldn't get added by AutoUpgrade either.

**Comment by yingopq** [comment]
> > @nikic I checked several other arch-32 situations with ForceEnableInt128, and they all use `alloca i128, align 16`. But they did not modify the target datalayout content with `i128:128`, should mips be consistent?
> 
> I think either way is fine here. Something to consider though, is that the AutoUpgrade for the DataLayout has to match. So if you don't add `i128:128` to the o32 DL, then it shouldn't get added by AutoUpgrade either.

OK, I decided to not add this.
Please help review whether there has other issue, thanks!

**Comment by yingopq** [comment]
Ping

**Comment by brad0** [comment]
@nikic

**Comment by nikic** [comment]
> > > @nikic I checked several other arch-32 situations with ForceEnableInt128, and they all use `alloca i128, align 16`. But they did not modify the target datalayout content with `i128:128`, should mips be consistent?
> > 
> > 
> > I think either way is fine here. Something to consider though, is that the AutoUpgrade for the DataLayout has to match. So if you don't add `i128:128` to the o32 DL, then it shouldn't get added by AutoUpgrade either.
> 
> OK, I decided to not add this. Please help review whether there has other issue, thanks!

In that case we also need to prevent the auto-upgrade if the o32 ABI is used. Otherwise auto-upgrade will produce the i128:128 data layout for o32, which directly targeting it would not.

**Comment by yingopq** [comment]
> > > > @nikic I checked several other arch-32 situations with ForceEnableInt128, and they all use `alloca i128, align 16`. But they did not modify the target datalayout content with `i128:128`, should mips be consistent?
> > > 
> > > 
> > > I think either way is fine here. Something to consider though, is that the AutoUpgrade for the DataLayout has to match. So if you don't add `i128:128` to the o32 DL, then it shouldn't get added by AutoUpgrade either.
> > 
> > 
> > OK, I decided to not add this. Please help review whether there has other issue, thanks!
> 
> In that case we also need to prevent the auto-upgrade if the o32 ABI is used. Otherwise auto-upgrade will produce the i128:128 data layout for o32, which directly targeting it would not.

OK, I updated the AutoUpgrade.cpp.

File: llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp [hunk 2]
```diff
@@ -75,6 +75,14 @@ TEST(DataLayoutUpgradeTest, ValidDataLayoutUpgrade) {
   EXPECT_EQ(UpgradeDataLayoutString("E-m:e-i64:64-n32:64-S128", "sparcv9"),
             "E-m:e-i64:64-i128:128-n32:64-S128");
 
+  // Check that MIPS64 targets add -i128:128.
+  EXPECT_EQ(UpgradeDataLayoutString(
+                "E-m:e-i8:8:32-i16:16:32-i64:64-n32:64-S128", "mips64"),
+            "E-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128");
+  EXPECT_EQ(UpgradeDataLayoutString(
+                "e-m:e-i8:8:32-i16:16:32-i64:64-n32:64-S128", "mips64el"),
+            "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128");
```
Comment (nikic) [inline llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp:84]: Can you please also test the non-upgrade for the o32 DL here?

**Review by nikic — APPROVED** [review]
LGTM, thanks!

File: llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp (hunk 2 above)
Comment (yingopq) [reply llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp:84]: OK, added.

**Comment by nikic** [comment]
@yingopq As you regularly do MIPS work, I'd recommend to request commit access, as described at https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access.

**Comment by yingopq** [comment]
> @yingopq As you regularly do MIPS work, I'd recommend to request commit access, as described at https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access.

Thanks for your recommendation， I would try to request commit access.

**Comment by yingopq** [comment]
> > @yingopq As you regularly do MIPS work, I'd recommend to request commit access, as described at https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access.
> 
> @nikic Thanks for your recommendation, I would try to request commit access.

Hi @nikic, do I need to send an email to apply myself, or can you apply for me? Thanks!

**Comment by brad0** [comment]
> Hi @nikic, do I need to send an email to apply myself, or can you apply for me? Thanks!

Does as it says in the first paragraph and send the email yourself.

**Comment by yingopq** [comment]
> > Hi @nikic, do I need to send an email to apply myself, or can you apply for me? Thanks!
> 
> Do as it says in the first paragraph and send the email yourself.

OK, thanks!

**Pull request merged**